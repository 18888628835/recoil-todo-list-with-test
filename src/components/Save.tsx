import React from 'react';

const Save = () => {
  const onSave = async () => {
    await fetch('/xxx');
  };
  return <button onClick={onSave}>Save</button>;
};

export default Save;
