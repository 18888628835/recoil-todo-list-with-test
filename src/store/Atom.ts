import { atom, selector } from 'recoil';

export interface TodoState {
  isComplete: boolean;
  id: number;
  text: string;
}

export const titleState = atom<string>({
  key: 'titleState',
  default: '',
});

export const todoListState = atom<TodoState[]>({
  key: 'todoListState',
  default: [],
});

export const todoListFilterState = atom({
  key: 'todoListFilterState',
  default: 'Show All',
});

export const filteredTodoListState = selector({
  key: 'filteredTodoListState',
  get: ({ get }) => {
    //这里的派生数据主要是根据 filter 值和 todoList 值进行过滤数据
    const filter = get(todoListFilterState);
    const list = get(todoListState);

    switch (filter) {
      case 'Show Completed':
        return list.filter((item) => item.isComplete);
      case 'Show Uncompleted':
        return list.filter((item) => !item.isComplete);
      default:
        return list;
    }
  },
});

export const todoListStatusState = selector({
  key: 'todoListStatusState',
  get: ({ get }) => {
    const todoList = get(todoListState);
    const totalNum = todoList.length;
    const totalCompletedNum = todoList.filter((item) => item.isComplete).length;
    const totalUncompletedNum = totalNum - totalCompletedNum;
    const percentCompleted = totalNum === 0 ? 0 : (totalCompletedNum / totalNum) * 100;

    return {
      totalNum,
      totalCompletedNum,
      totalUncompletedNum,
      percentCompleted,
    };
  },
});
