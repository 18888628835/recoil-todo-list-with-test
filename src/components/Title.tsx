import React, { useEffect } from 'react';
import { useRecoilState } from 'recoil';
import { titleState } from '../store/Atom';

const getTitleData = () => {
  const api = 'https://api.github.com/users/18888628835';
  return fetch(api)
    .then((res) => res.json())
    .catch(() => 'title from fetch');
};
const Title = () => {
  const [title, setTitle] = useRecoilState(titleState);

  useEffect(() => {
    async function getAndSetTitle() {
      const res = await getTitleData();
      setTitle(res.login);
    }
    getAndSetTitle();
  }, [setTitle]);

  return <div>{title}</div>;
};

export default Title;
