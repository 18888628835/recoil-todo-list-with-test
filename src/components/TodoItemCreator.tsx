import { ChangeEvent, useState } from 'react';
import { useSetRecoilState } from 'recoil';
import { todoListState } from '../store/Atom';
// utility for creating unique Id
let id = 0;
function getId() {
  return id++;
}
//新增数据
function TodoItemCreator() {
  const [inputValue, setInputValue] = useState('');
  //这里需要读写数据，所以用useSetRecoilState
  const setTodoList = useSetRecoilState(todoListState);

  const addItem = () => {
    setTodoList((oldTodoList) => [
      ...oldTodoList,
      {
        id: getId(),
        text: inputValue,
        isComplete: false,
      },
    ]);
    setInputValue('');
  };

  const onChange = (e: ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value);
  };

  return (
    <div>
      <input type="text" value={inputValue} onChange={onChange} />
      <button onClick={addItem}>Add</button>
    </div>
  );
}

export default TodoItemCreator;
