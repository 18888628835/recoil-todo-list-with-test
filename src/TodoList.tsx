import TodoItemCreator from './components/TodoItemCreator';
import TodoListFilters from './components/TodoListFilters';
import TodoListStatus from './components/TodoListStatus';
import TodoItemsRender from './components/TodoItemsRender';
import Title from './components/Title';
import Save from './components/Save';

export default function TodoList() {
  return (
    <div>
      <Title />
      <TodoItemCreator />
      <TodoItemsRender />
      <TodoListFilters />
      <TodoListStatus />
      <Save />
    </div>
  );
}
