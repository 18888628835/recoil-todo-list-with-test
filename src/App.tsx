import { RecoilRoot } from 'recoil';
import TodoList from './TodoList';
import './App.css';

function App() {
  return (
    <div className="App">
      <RecoilRoot>
        <TodoList />
      </RecoilRoot>
    </div>
  );
}

export default App;
