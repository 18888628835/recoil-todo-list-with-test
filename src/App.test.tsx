import { act, render, screen, waitFor, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useEffect } from 'react';
import { RecoilRoot, useRecoilValue } from 'recoil';
import App from './App';
import Save from './components/Save';
import Title from './components/Title';
import TodoItemCreator from './components/TodoItemCreator';
import { todoListState } from './store/Atom';
import TodoList from './TodoList';

/* 
<input type="text" value={inputValue} onChange={onChange} />;
*/
it('input can be typed', () => {
  render(<TodoItemCreator />);
  const inputEl = screen.getByRole('textbox');
  userEvent.type(inputEl, 'type some text ');
  expect(inputEl.getAttribute('value')).toBe('type some text ');
  userEvent.type(inputEl, 'continue');
  expect(inputEl.getAttribute('value')).not.toBe('continue');
  expect(screen.getByDisplayValue('type some text continue')).toBeInTheDocument();
});

/**
  const options = [
    { label: 'All', value: 'Show All' },
    { label: 'Completed', value: 'Show Completed' },
    { label: 'Uncompleted', value: 'Show Uncompleted' },
  ]; 
  <select>
  {options.map(({ label, value }, index) => (
    <option key={index} value={value} data-testid="select-option">
      {label}
    </option>
  ))}
</select> 
 */

it.skip('select options', () => {
  render(<App />);
  const options = screen.getAllByTestId('select-option');
  expect(options[0].getAttribute('value')).toBe('Show All');
  expect(options[1].getAttribute('value')).toBe('Show Completed');
  expect(options[2].getAttribute('value')).toBe('Show Uncompleted');
});
/**
 useEffect(() => {
    const dataFromDB = [
      { label: 'All', value: 'Show All', disabled: false },
      { label: 'Completed', value: 'Show Completed', disabled: false },
      { label: 'Uncompleted', value: 'Show Uncompleted', disabled: true },
    ];
    Promise.resolve(dataFromDB).then((res) => setOptions([...res]));
  }, []);
 */
it('async get select options', async () => {
  render(<App />);
  const options = await screen.findAllByTestId('select-option');
  expect(options[0].getAttribute('value')).toBe('Show All');
  expect(options[1].getAttribute('value')).toBe('Show Completed');
  expect(options[2].getAttribute('value')).toBe('Show Uncompleted');
});

it('default select-option', async () => {
  render(<App />);
  const options = await screen.findAllByTestId('select-option');
  expect((options[0] as HTMLOptionElement).selected).toBeTruthy();
  expect((options[1] as HTMLOptionElement).selected).toBeFalsy();
  expect((options[2] as HTMLOptionElement).selected).toBeFalsy();
});

it('select-option selected', async () => {
  render(<App />);
  const selectEl = await screen.findByRole('combobox');
  const options = await screen.findAllByTestId('select-option');
  userEvent.selectOptions(selectEl, options[1]);
  expect((options[0] as HTMLOptionElement).selected).toBeFalsy();
  expect((options[1] as HTMLOptionElement).selected).toBeTruthy();
  expect((options[2] as HTMLOptionElement).selected).toBeFalsy();
});

it('one select-option is disabled', async () => {
  render(<App />);
  const options = await screen.findAllByTestId('select-option');
  // The following code will report an error
  // userEvent.selectOptions(selectEl, options[3]);
  expect(options[2]).toBeDisabled();
});

/**
    <ul>
      <li>Total items: {totalNum}</li>
      <li>Items completed: {totalCompletedNum}</li>
      <li>Items not completed: {totalUncompletedNum}</li>
      <li>Percent completed: {formattedPercentCompleted}</li>
    </ul>
 */
it('should render list of 4 list-items', async () => {
  render(<App />);
  await waitFor(() => {
    const list = screen.getByRole('list');
    const { getAllByRole } = within(list);
    const items = getAllByRole('listitem');
    expect(items.length).toBe(4);
  });
});

it('match snapshot', async () => {
  const { asFragment } = render(<App />);
  await screen.findAllByTestId('select-option');
  expect(asFragment()).toMatchSnapshot();
});

it('add Item', async () => {
  render(<App />);
  const inputEl = screen.getByRole('textbox');
  const buttonEl = screen.getByRole('button', { name: /add/i });
  userEvent.type(inputEl, 'item1');
  userEvent.click(buttonEl);
  await act(async () => Promise.resolve());
  expect(await screen.findByDisplayValue('item1')).toBeInTheDocument();
  expect(screen.queryByDisplayValue('item2')).toBeNull();
  expect(screen.queryByDisplayValue('item3')).not.toBeInTheDocument();
});

it('delete Item', async () => {
  render(<App />);
  const inputEl = screen.getByRole('textbox');
  const buttonEl = screen.getByRole('button', { name: /add/i });
  userEvent.type(inputEl, 'item1');
  userEvent.click(buttonEl);
  await act(() => Promise.resolve());
  const todoItem = screen.getByTestId('todo-item-0');
  const { findByRole } = within(todoItem);
  const deleteEl = await findByRole('button', { name: /x/i });
  userEvent.click(deleteEl);
  expect(screen.queryByDisplayValue('item1')).toBeNull();
  // expect(await screen.findByDisplayValue('item1')).toBeInTheDocument();
});

describe('the recoil todoListState should', () => {
  // @ts-ignore
  const RecoilObserver = ({ state, onChange }) => {
    const value = useRecoilValue(state);
    useEffect(() => onChange(value), [onChange, value]);
    return null;
  };
  it('be change when user click add button and x button', async () => {
    const onChange = jest.fn();
    render(
      <RecoilRoot>
        <RecoilObserver state={todoListState} onChange={onChange} />
        <TodoList />
      </RecoilRoot>
    );
    const inputEl = await screen.findByRole('textbox');
    const buttonEl = await screen.findByRole('button', { name: /add/i });
    // the following code will report error: act(...)
    //const inputEl = screen.getByRole('textbox');
    //const buttonEl = screen.getByRole('button', { name: /add/i });
    userEvent.type(inputEl, 'item1');
    userEvent.click(buttonEl);
    expect(onChange).toHaveBeenCalledTimes(2);
    expect(onChange).toHaveBeenCalledWith([]);
    expect(onChange).toHaveBeenCalledWith([{ id: expect.any(Number), isComplete: false, text: 'item1' }]);

    const todoItem = screen.getByTestId('todo-item-0');
    const { findByRole } = within(todoItem);
    const deleteEl = await findByRole('button', { name: /x/i });
    userEvent.click(deleteEl);
    expect(onChange).toHaveBeenCalledWith([]);
  });

  it('async get data from backend', async () => {
    const mockState = { login: 'test title' };
    global.fetch = jest.fn(() =>
      Promise.resolve({
        json: () => Promise.resolve(mockState),
      })
    ) as jest.Mock;

    render(
      <RecoilRoot>
        <Title />
      </RecoilRoot>
    );
    expect(await screen.findByText(mockState.login)).toBeInTheDocument();
  });

  it('save data', async () => {
    global.fetch = jest.fn() as jest.Mock;
    render(
      <RecoilRoot>
        <Save />
      </RecoilRoot>
    );
    const buttonEl = screen.getByRole('button', { name: /Save/i });
    userEvent.click(buttonEl);
    expect(global.fetch).toBeCalledTimes(1);
  });
});
