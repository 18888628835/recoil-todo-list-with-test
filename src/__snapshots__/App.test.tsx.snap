// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div>
      <div />
      <div>
        <input
          type="text"
          value=""
        />
        <button>
          Add
        </button>
      </div>
      <div />
      Filter:
      <select>
        <option
          data-testid="select-option"
          value="Show All"
        >
          All
        </option>
        <option
          data-testid="select-option"
          value="Show Completed"
        >
          Completed
        </option>
        <option
          data-testid="select-option"
          disabled=""
          value="Show Uncompleted"
        >
          Uncompleted
        </option>
      </select>
      <ul>
        <li>
          Total items: 0
        </li>
        <li>
          Items completed: 0
        </li>
        <li>
          Items not completed: 0
        </li>
        <li>
          Percent completed: 0
        </li>
      </ul>
      <button>
        Save
      </button>
    </div>
  </div>
</DocumentFragment>
`;
